### ISSUE TITLE
Public Search: Guardrails for Advanced Order and Opinion Search Queries

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8093]
As a DAWSON administrator, so that I can maintain system stability once public Order search is enabled, I need to be able to increase or decrease the rate at which all (internal and public) advanced order and opinion queries are submitted through the UI.  Both of these types of searches currently utilize a join in their queries and can have an outsized impact on cluster performance (compared to case search).

Before going live with public Order search, we would like to ensure that the Court can respond to heavy or unexpected traffic causing load on Elastic Search by scaling up or down the allowed number of concurrent calls to the advanced Order search and Opinion search by ALL users both internal and public. 

Once the Court has determined appropriate thresholds for triggering these 'guardrails' based on CPU usage of Elastic Search, the system should adjust traffic as appropriate. This will allow us to make sure that the overall system performance isn't affected by bad actors or sudden load. 

Efforts in this story will be removing the slow-down limiter API, but the user-specific limiter should remain in place.

## Pre-Conditions

- [x] Court has determined the maximum concurrent rate of advanced document searches per minute: **5/sec; 300/min should be the default**

## Acceptance Criteria
* Court Tech Lead can increase or decrease the maximum concurrent rate of advanced document searches per minute
* Court Tech Lead can do this without having to deploy new code
* Adjusting the maximum concurrent rate affects traffic both internally and externally
* Solution can be implemented and tested in Court pre-production environments and Flexion Staging environment
* Server responds with HTTP 429/Too Many Requests when the limit has been reached
* UI reflects appropriate messaging (EDIT: offloaded to ustaxcourt/ef-cms#8107 )
* Documentation for how to use this is in the docs folder

## Notes
* Possible companion story: Client handles 429 response with a friendly User Message to inform the user to wait a few and try again (pretty sure we have some of this messaging already available)

## Tasks
Apply limiter to:
- [ ] internal order search
- [ ] public order search
- [ ] internal opinion search
- [ ] public opinion search
- [ ] remove slow-down limiter from internal middleware
- [ ] verify that public UI correctly handles HTTP 429 codes
- [ ] verify that internal UI correctly handles HTTP 429 codes

## Story Definition of Ready (Created on 9/26/21)
The following criteria must be met in order for the user story to be ready for grooming.
The user story must: 
- [ ]  Be immediately actionable for the development team
- [ ]  Be focused on users
- [ ]  Include a narrative. “As a ____, so that I can ______, I need _____.”
- [ ]  Include user-centric acceptance criteria that expresses the users' needs
- [ ]  Include user-centric testing scenarios
- [ ]  Include preconditions (story dependencies)
- [ ]  Not include design or technical solutions
- [ ]  Include non-standard testing, deployment, documentation, browser, mobile or security requirements, as well as any requirements for 3rd-party integrations

Process: Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [x]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios to meet all acceptance criteria have been written
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] Deployed to any Flexion environment for UX Review
 - [x] Deployed to the Court's migration environment

**Comment by mariahkannenberg** [comment]
Needs More Info - Write test cases

**Comment by cholly75** [comment]
Test Cases:

1) Set allowable limit of calls to X. Generate load greater than X. Calls are not passed through to ES, UI reflects same.
2) Set allowable limit of calls to X. Generate load less than X. Calls are passed through to ES, search functions as normal.
3) Set allowable limit of calls to X. Generate load less than X. Search functions as normal. Change value of X to a lower value. Calls stop being passed through to ES, UI reflects the same.
4) Set allowable limit of calls to X. Generate load higher than X. Calls are not passed through to ES, UI reflects the same. Change value of X to a higher value. Search functions as normal.

**Comment by kkoskelin** [comment]
Please note:  There are several integration tests currently disabled relating to public search for orders.  Once this feature is enabled, those tests should be turned back on.  See linting warnings for more info.

**Comment by codyseibert** [comment]
Note for deployments:

The advancedQueryLimiter created during this story work is reading configuration from the `efcms-deploy-ENV` table and expects a record to be in the following format:

```
{
  "pk": "document-search-limiter-configuration",
  "sk": "document-search-limiter-configuration",
  "maxInvocations": 5,
  "windowTime": 1000
}
```

`windowTime` is how long to wait (in milliseconds) until the limiter should reset it's internal timer
`maxInvocations` is how many queries we want to allow in the allowed windowTime.

**Comment by mariahkannenberg** [comment]
@TomElliottFlexion What's the blocker on this one?

**Comment by TomElliottFlexion** [comment]
Just waiting for Batch 11 to finish the move to staging, before we can get this on mig

**Comment by mmarcotte** [comment]
Paring with @TomElliottFlexion , @codyseibert , and @Absolutestunna we did some performance testing on this story and everything looked great! 

However, we did observe one thing that still needed to happen, and that is to reintroduce the IP Limiter; this will prevent bad actors from triggering the throttle and interfering with other people's ability to search once this goes live to the public. 

Here are the results (observe the three spikes at the end of the timeline):
<img width="1055" alt="Screen Shot 2021-10-22 at 11 39 53 AM" src="https://user-images.githubusercontent.com/5023502/138491496-2e1c90bc-90ff-42a3-ad99-1f1de4635094.png">

The last three load tests were all performed at a rate of 20 req/sec. We had the throttle set to 3, 5, and then 8 requests per second. We observed a proportionate amount of 429 and 200 requests. None of our requests timed out.

### test 1

* throttle: 3 per sec, 
* duration 5 minutes,
* arrivateRate: 20 (20 req per sec)

```
    "requestsCompleted": 6000,
    "latency": {
      "min": 60,
      "max": 4071,
      "median": 132,
      "p95": 762.5,
      "p99": 1979.5
    },
    "rps": {
      "count": 6000,
      "mean": 19.94
    },
    ...
    "errors": {},
    "codes": {
      "200": 1139,
      "429": 4861
    },
```

### test 2

* throttle: 5 per sec
* duration: 5 minutes
* arrivalRate: 20 req per sec

```
    "requestsCompleted": 6000,
    "latency": {
      "min": 57,
      "max": 3541,
      "median": 134,
      "p95": 897,
      "p99": 1256
    },
    "rps": {
      "count": 6000,
      "mean": 19.88
    },
    ...
    "errors": {},
    "codes": {
      "200": 1618,
      "429": 4382
    },
```

### Test 3

* throttle: 8 per sec
* duration: 5 minutes
* arrivalRate: 20 per sec

```
    "requestsCompleted": 6000,
    "latency": {
      "min": 56,
      "max": 7681,
      "median": 150,
      "p95": 1324,
      "p99": 3172.5
    },
    "rps": {
      "count": 6000,
      "mean": 19.94
    },
    ...
    "errors": {},
    "codes": {
      "200": 2358,
      "429": 3642
    },
```

**Comment by cholly75** [comment]
@mmarcotte Do you have the story # for the IP limiter? I can make a new story to re-introduce that.
